### PR TITLE
feat: autoPipelining

### PR DESCRIPTION
**Pull request by gurgunday**
Not eager to explain why it's structured this way... I'm open to suggestions

I could do a much better design if we added support for async `incr` functions and made more breaking changes, which I have done in another branch, but one step at a time... 🤣

If anyone finds this messy, I can just remove the old version but we would really have to recommend users to enable `autoPipelining`

### CODE REVIEW

**Review by Uzlopak — CHANGES_REQUESTED** [review]
I add a blocker.

**Comment by gurgunday** [comment]
@Uzlopak what do you think? This is without using Lua and should work, having two incr functions is not great

**Comment by Uzlopak** [comment]
I am currently bisecting node. I will come back when I finished with node

**Review by gurgunday — COMMENTED** [review]
I don’t find this code pleasing to look at

.pipeline does the same thing without mixing async, await, and callbacks

I’ll let this stay open for a bit and close it

File: store/RedisStore.js [hunk 1]
```diff
@@ -5,46 +5,70 @@ function RedisStore (redis, timeWindow, continueExceeding, key) {
...
 
-RedisStore.prototype.incr = function (ip, cb, max) {
+RedisStore.prototype.incrAutoPipeline = async function (ip, cb, max) {
   const key = this.key + ip
+  const [current, ttl] = await Promise.all([this.redis.incr(key), this.redis.pttl(key)])
+
+  if (ttl === -1) {
+    // Item just got created
+    this.redis.pexpire(key, this.timeWindow).catch()
+    cb(null, { current: 1, ttl: this.timeWindow })
```
Comment (mcollina) [inline store/RedisStore.js:23]: I think we need to add these as callback to pexpire to avoid test issues

Comment (gurgunday) [reply store/RedisStore.js:23]: Yeah I think that’s an improvement, just wanted to keep the behavior the same - it was like this before even with callbacks

Comment (gurgunday) [reply store/RedisStore.js:23]: What do you think about making the functions async? Without making them async, I can’t group them together without .pipeline

Comment (mcollina) [reply store/RedisStore.js:23]: that would be a semver-major.

Comment (mcollina) [reply store/RedisStore.js:23]: +1 for me

Comment (Uzlopak) [reply store/RedisStore.js:23]: I really would prefer a customCommand which runs the db related stuff on redis over more fiddling in the business code.

**Pull request closed without merging**